src/legacy/core_plugins/dashboard_embeddable/public/actions/expand_panel_action.tsx

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { EuiIcon } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import {
+  Action,
+  actionRegistry,
+  IEmbeddable,
+  CONTEXT_MENU_TRIGGER,
+  triggerRegistry,
+  ActionContext,
+  ExecuteActionContext,
+} from 'plugins/embeddable_api/index';
+import React from 'react';
+import { IncompatibleActionError } from 'plugins/embeddable_api/actions';
+import { DASHBOARD_CONTAINER_TYPE, DashboardContainer } from '../embeddable';
+
+export const EXPAND_PANEL_ACTION = 'togglePanel';
+
+function isDashboard(
+  embeddable: IEmbeddable | DashboardContainer
+): embeddable is DashboardContainer {
+  return (embeddable as DashboardContainer).type === DASHBOARD_CONTAINER_TYPE;
+}
+
+function isExpanded(embeddable: IEmbeddable) {
+  if (!embeddable.parent || !isDashboard(embeddable.parent)) {
+    throw new IncompatibleActionError();
+  }
+
+  return embeddable.id === embeddable.parent.getInput().expandedPanelId;
+}
+
+export class ExpandPanelAction extends Action {
+  public readonly type = EXPAND_PANEL_ACTION;
+
+  constructor() {
+    super(EXPAND_PANEL_ACTION);
+    this.order = 7;
+  }
+
+  public getDisplayName({ embeddable }: ActionContext) {
+    if (!embeddable.parent || !isDashboard(embeddable.parent)) {
+      throw new IncompatibleActionError();
+    }
+
+    return isExpanded(embeddable)
+      ? i18n.translate('kbn.embeddable.actions.toggleExpandPanel.expandedDisplayName', {
+          defaultMessage: 'Minimize',
+        })
+      : i18n.translate('kbn.embeddable.actions.toggleExpandPanel.notExpandedDisplayName', {
+          defaultMessage: 'Full screen',
+        });
+  }
+
+  public getIcon({ embeddable }: ActionContext) {
+    if (!embeddable.parent || !isDashboard(embeddable.parent)) {
+      throw new IncompatibleActionError();
+    }
+    return <EuiIcon type={isExpanded(embeddable) ? 'expand' : 'expand'} />;
```

Comment:
```
  public getIcon(context: ActionContext): EuiContextMenuItemIcon | undefined {
    return undefined;
  }
```

type at least in base Action class is `EuiContextMenuItemIcon`, though I could see string being appropriate too (or either).  Maybe elsewhere it has a string type?

Benefit of requesting EuiContextMenuItemIcon is that users can create their own icons (yes, still restricted to react), but we know it'll look appropriate it the EuiContext Menu. 

I'm okay with that coupling here, and it's optional. The embeddable rendering component can be built in any framework, we want it to be really flexible.  Icon though, we want to have control over the look and feel of it. So we are using react, if you want an icon, use react (or, maybe at that point, we say, use a string otherwise... though I don't think that functionality exists yet).